params/hooks_libevm.go

```diff
@@ -52,48 +57,44 @@ func (r RulesExtra) PrecompileOverride(addr common.Address) (libevm.PrecompiledC
 		return nil, false
 	}
 
-	return libevmContract{module.Contract}, true
-}
-
-// XXX: This is a hack since we need the suppliedGas
-// to determine the gas cost of the precompile
-// also evm.interpreter.ReadOnly
-type libevmContract struct {
-	contract.StatefulPrecompiledContract
-}
-
-func (l libevmContract) RunExtra(
-	chainConfig *gethparams.ChainConfig,
-	blockNumber *big.Int, blockTime uint64, predicateResults libevm.PredicateResults,
-	state libevm.StateDB, _ *gethparams.Rules, caller, self common.Address, input []byte, suppliedGas uint64, readOnly bool) ([]byte, uint64, error) {
-	accessableState := accessableState{
-		StateDB:     state,
-		chainConfig: chainConfig,
-		blockContext: &BlockContext{
-			number:           blockNumber,
-			time:             blockTime,
-			predicateResults: predicateResults,
-		}}
-	return l.StatefulPrecompiledContract.Run(accessableState, caller, self, input, suppliedGas, readOnly)
-}
-
-func (libevmContract) Run(input []byte) ([]byte, error) {
-	panic("implement me")
-}
-
-func (libevmContract) RequiredGas(input []byte) uint64 {
-	panic("implement me")
+	precompile := func(env vm.PrecompileEnvironment, input []byte, suppliedGas uint64) ([]byte, uint64, error) {
+		header, err := env.BlockHeader()
+		if err != nil {
+			panic(err) // Should never happen
+		}
+		predicateResults, err := PredicateParser(header.Extra)
+		if err != nil {
+			panic(err) // Should never happen, because predicates are parsed in NewEVMBlockContext.
+		}
+		// XXX: this should be moved to the precompiles
+		var state libevm.StateReader
+		if env.ReadOnly() {
+			state = env.ReadOnlyState()
+		} else {
+			state = env.StateDB()
+		}
+		accessableState := accessableState{
+			StateReader: state,
+			chainConfig: GetRulesExtra(env.Rules()).chainConfig,
+			blockContext: &BlockContext{
+				number:           env.BlockNumber(),
+				time:             env.BlockTime(),
+				predicateResults: predicateResults,
+			}}
+		return module.Contract.Run(accessableState, env.Addresses().Caller, env.Addresses().Self, input, suppliedGas, env.ReadOnly())
+	}
+	return vm.NewStatefulPrecompile(precompile), true
 }
 
 type accessableState struct {
-	libevm.StateDB
+	libevm.StateReader
 	chainConfig  *gethparams.ChainConfig
 	blockContext *BlockContext
 }
 
 func (a accessableState) GetStateDB() contract.StateDB {
 	// XXX: Whoa, this is a hack
-	return a.StateDB.(contract.StateDB)
+	return a.StateReader.(contract.StateDB)
```

Comment:
Is this just to make it compliant with an existing interface?

To guarantee that it's safe, without using the `, ok` idiom you could:

```go
var (
    sr libevm.StateReader = a.StateReader
    _ contract.StateDB = (libevm.StateReader)(nil) // prove that the next line won't panic
)
return sr.(contract.StateDB)
```

It's gross, but at least it's temporary!